# PyCity Schools Analysis


In [42]:
# Dependencies and Setup
import pandas as pd

# Define the file paths and Read School & Student Data File
school_data = pd.read_csv("Resources/schools_complete.csv")
student_data = pd.read_csv("Resources/students_complete.csv")

# Merge the data into a single dataset
complete_data = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

# Display the merged dataset
complete_data.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Perform the necessary calculations and then create a high-level snapshot of the district's key metrics in a DataFrame.

* Include the following:

* Total number of unique schools

* Total students

* Total budget

* Average math score

* Average reading score

* % passing math (the percentage of students who passed math)

* % passing reading (the percentage of students who passed reading)

* % overall passing (the percentage of students who passed math AND reading)

In [43]:
# Calculate the total number of unique schools
unique_schools = complete_data["school_name"].nunique()
unique_schools

15

In [44]:
# Calculate the total number of students
total_students = complete_data["school_name"].count()
total_students

39170

In [45]:
# Calculate the total budget
total_budgets = school_df["budget"].sum()
total_budgets

24649428

In [46]:
# Calculate the average (mean) math score
avg_math_score = complete_data["math_score"].mean()
avg_math_score

78.98537145774827

In [47]:
# Calculate the average (mean) reading score
avg_reading_score = complete_data["reading_score"].mean()
avg_reading_score

81.87784018381414

In [48]:
# Use the following to calculate the percentage of students who passed math (math scores greather than or equal to 70)
math_count_pass = complete_data[complete_data["math_score"] >= 70].count()
percentage_math_pass = (math_count_pass / total_students) * 100
percentage_math_pass = percentage_math_pass["math_score"]
percentage_math_pass

74.9808526933878

In [49]:
# Calculate the percentage of students who passed reading (hint: look at how the math percentage was calculated)  
reading_count_pass = complete_data[complete_data["reading_score"] >= 70].count()
percentage_reading_pass = (reading_count_pass / total_students) * 100
percentage_reading_pass = percentage_reading_pass["reading_score"]
percentage_reading_pass

85.80546336482001

In [50]:
# Use the following to calculate the percentage of students that passed math and reading
overall_passing_count = complete_data[(complete_data["reading_score"] >= 70) & (complete_data["math_score"] >= 70)].count()
overall_passing_percentage = (overall_passing_count / total_students) * 100
overall_passing_percentage = overall_passing_percentage["math_score"]
overall_passing_percentage

65.17232575950983

In [51]:
# Create a high-level snapshot of the district's key metrics in a DataFrame
district_summary = {
    'Total Schools': [unique_schools],
    'Total Students': [total_students],
    'Total Budget': [total_budgets],
    'Average Math Score': [avg_math_score],
    'Average Reading Score' : [avg_reading_score],
    '% Passing Math': [percentage_math_pass],
    '% Passing Reading': [percentage_reading_pass],
    '% Overall Passing': [overall_passing_percentage]
}

# Create a DataFrame from the result dictionary
district_summary_df = pd.DataFrame(district_summary)

format_mapping = {
    'Total Students': '{:,}',
    'Total Budget': '${:,.2f}',
    'Average Math Score': '{:.4f}',
    'Average Reading Score': '{:.4f}',
    '% Passing Math': '{:.4f}',
    '% Passing Reading': '{:.4f}',
    '% Overall Passing': '{:.4f}'
}

# Apply formatting to the specified columns using the format_mapping dictionary
for column, fmt in format_mapping.items():
    district_summary_df[column] = district_summary_df[column].map(fmt.format)

# Set the index name to empty string
district_summary_df.index.name = ''

# Display the data
district_summary_df

Total Schools Total Students    Total Budget Average Math Score  \
                                                                     
0             15         39,170  $24,649,428.00            78.9854   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
                                                                            
0               81.8778        74.9809           85.8055           65.1723

# School Summary
* Use the code provided to select the school type (2 points)
* Calculate the total student count (2 points)
* Use the code provided to calculate the per capita spending (2 points)
* Calculate the average test scores (2 points)
* Calculate the number of schools with math scores of 70 or higher (2 points)
* Calculate the number of schools with reading scores of 70 or higher (2 points)
* Use the provided code to calculate the schools that passed both math and reading with scores of 70 or higher (2 points)
* Use the provided code to calculate the passing rates (2 points)
* Create a new DataFrame for the above calculations called per_school_summary (4 points)

In [78]:
# Group by school name
school_group = school_data_complete.groupby('school_name')

In [79]:
# School type by school name
school_type = school_data.set_index('school_name')['type']

In [80]:
# Calculate total students
total_students = school_group['Student ID'].count()

In [81]:
# Total school budget
total_school_budget = school_data.set_index('school_name')['budget']

In [82]:
# Per student budget
budget_per_student = total_school_budget / school_data.set_index('school_name')['size']

In [83]:
# Average Math Score
average_math_score = school_group['math_score'].mean()

In [84]:
# Average Reading Score
average_reading_score = school_group['reading_score'].mean()

In [85]:
# Percentage of students passing math
percent_passing_math = school_data_complete[school_data_complete['math_score'] >= 70].groupby('school_name')['Student ID'].count() / total_students * 100

In [86]:
# Percentage of students passing reading
percent_passing_reading = school_data_complete[school_data_complete['reading_score'] >= 70].groupby('school_name')['Student ID'].count() / total_students * 100

In [87]:
# Percentage of students passing both math and reading
percent_overall_passing = school_data_complete[(school_data_complete['reading_score'] >= 70) 
                                               & (school_data_complete['math_score'] >= 70)].groupby('school_name')['Student ID'].count() / total_students * 100

In [88]:
# Create a DataFrame with the school summary
per_school_summary = pd.DataFrame({
    "School Type": school_type,
    "Total Students": total_students,
    "Total School Budget": total_school_budget,
    "Per Student Budget": budget_per_student,
    "Average Math Score": average_math_score,
    "Average Reading Score": average_reading_score,
    "% Passing Math": percent_passing_math,
    "% Passing Reading": percent_passing_reading,
    "% Overall Passing": percent_overall_passing
})

# Formatting
per_school_summary['Total Students'] = per_school_summary['Total Students'].map('{:,}'.format)
per_school_summary['Total School Budget'] = per_school_summary['Total School Budget'].map('${:,.2f}'.format)

# Display the school summary
per_school_summary

School Type Total Students Total School Budget  \
school_name                                                            
Bailey High School       District          4,976       $3,124,928.00   
Cabrera High School       Charter          1,858       $1,081,356.00   
Figueroa High School     District          2,949       $1,884,411.00   
Ford High School         District          2,739       $1,763,916.00   
Griffin High School       Charter          1,468         $917,500.00   
Hernandez High School    District          4,635       $3,022,020.00   
Holden High School        Charter            427         $248,087.00   
Huang High School        District          2,917       $1,910,635.00   
Johnson High School      District          4,761       $3,094,650.00   
Pena High School          Charter            962         $585,858.00   
Rodriguez High School    District          3,999       $2,547,363.00   
Shelton High School       Charter          1,761       $1,056,600.00   
Thomas High School        Charter          1,635       $1,043,130.00   
Wilson High School        Charter          2,283       $1,319,574.00   
Wright High School        Charter          1,800       $1,049,400.00   

                       Per Student Budget  Average Math Score  \
school_name                                                     
Bailey High School                  628.0           77.048432   
Cabrera High School                 582.0           83.061895   
Figueroa High School                639.0           76.711767   
Ford High School                    644.0           77.102592   
Griffin High School                 625.0           83.351499   
Hernandez High School               652.0           77.289752   
Holden High School                  581.0           83.803279   
Huang High School                   655.0           76.629414   
Johnson High School                 650.0           77.072464   
Pena High School                    609.0           83.839917   
Rodriguez High School               637.0           76.842711   
Shelton High School                 600.0           83.359455   
Thomas High School                  638.0           83.418349   
Wilson High School                  578.0           83.274201   
Wright High School                  583.0           83.682222   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School             81.933280          54.642283  
Cabrera High School            97.039828          91.334769  
Figueroa High School           80.739234          53.204476  
Ford High School               79.299014          54.289887  
Griffin High School            97.138965          90.599455  
Hernandez High School          80.862999          53.527508  
Holden High School             96.252927          89.227166  
Huang High School              81.316421    

## Highest-Performing Schools by Percentage of Overall Passing (5 points)

* Sort the schools by % Overall Passing in descending order (2 points)
* Save the results to a DataFrame called top_schools (2 points)
* Display the first 5 rows (1 point)

In [89]:
# Sort the schools by `% Overall Passing` in descending order and display the top 5 rows.
top_schools = per_school_summary.sort_values('% Overall Passing', ascending = False)
top_schools.head(5)

School Type Total Students Total School Budget  \
school_name                                                          
Cabrera High School     Charter          1,858       $1,081,356.00   
Thomas High School      Charter          1,635       $1,043,130.00   
Griffin High School     Charter          1,468         $917,500.00   
Wilson High School      Charter          2,283       $1,319,574.00   
Pena High School        Charter            962         $585,858.00   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   
Pena High School                  609.0           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
school_name                                                                     
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
school_name                             
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

## Bottom Performing Schools (By % Overall Passing)
* Sort the schools by % Overall Passing in ascending order (2 points)
* Save the results to a DataFrame called bottom_schools (2 points)
* Display the first 5 rows (1 point)

In [90]:
# Sort the schools by `% Overall Passing` in ascending order and display the top 5 rows.
bottom_schools = per_school_summary.sort_values('% Overall Passing', ascending = True)
bottom_schools.head(5)

School Type Total Students Total School Budget  \
school_name                                                            
Rodriguez High School    District          3,999       $2,547,363.00   
Figueroa High School     District          2,949       $1,884,411.00   
Huang High School        District          2,917       $1,910,635.00   
Hernandez High School    District          4,635       $3,022,020.00   
Johnson High School      District          4,761       $3,094,650.00   

                       Per Student Budget  Average Math Score  \
school_name                                                     
Rodriguez High School               637.0           76.842711   
Figueroa High School                639.0           76.711767   
Huang High School                   655.0           76.629414   
Hernandez High School               652.0           77.289752   
Johnson High School                 650.0           77.072464   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

## Math Scores by Grade
* Use the code provided to separate the data by grade (1 points)
* Group by "school_name" and take the mean of each (4 points)
* Use the code to select only the math_score (1 points)
* Combine each of the scores above into single DataFrame called math_scores_by_grade (4 points)

In [91]:
# Use the code provided to separate the data by grade
ninth_graders = school_data_complete[school_data_complete["grade"] == "9th"]
tenth_graders = school_data_complete[school_data_complete["grade"] == "10th"]
eleventh_graders = school_data_complete[school_data_complete["grade"] == "11th"]
twelfth_graders = school_data_complete[school_data_complete["grade"] == "12th"]

# Group by `school_name` and take the mean of the `math_score` column for each grade
ninth_grade_math_scores = ninth_graders.groupby("school_name")["math_score"].mean()
tenth_grade_math_scores = tenth_graders.groupby("school_name")["math_score"].mean()
eleventh_grade_math_scores = eleventh_graders.groupby("school_name")["math_score"].mean()
twelfth_grade_math_scores = twelfth_graders.groupby("school_name")["math_score"].mean()

# Combine each of the scores above into a single DataFrame called `math_scores_by_grade`
math_scores_by_grade = pd.DataFrame({
    "9th Grade": ninth_grade_math_scores,
    "10th Grade": tenth_grade_math_scores,
    "11th Grade": eleventh_grade_math_scores,
    "12th Grade": twelfth_grade_math_scores
})

# Minor data wrangling
math_scores_by_grade.index.name = None

# Display the DataFrame
math_scores_by_grade


9th Grade  10th Grade  11th Grade  12th Grade
Bailey High School     77.083676   76.996772   77.515588   76.492218
Cabrera High School    83.094697   83.154506   82.765560   83.277487
Figueroa High School   76.403037   76.539974   76.884344   77.151369
Ford High School       77.361345   77.672316   76.918058   76.179963
Griffin High School    82.044010   84.229064   83.842105   83.356164
Hernandez High School  77.438495   77.337408   77.136029   77.186567
Holden High School     83.787402   83.429825   85.000000   82.855422
Huang High School      77.027251   75.908735   76.446602   77.225641
Johnson High School    77.187857   76.691117   77.491653   76.863248
Pena High School       83.625455   83.372000   84.328125   84.121547
Rodriguez High School  76.859966   76.612500   76.395626   77.690748
Shelton High School    83.420755   82.917411   83.383495   83.778976
Thomas High School     83.590022   83.087886   83.498795   83.497041
Wilson High School     83.085578   83.724422   83.195326   83.035794
Wright High School     83.264706   84.010288   83.836782   83.644986

## Reading Scores by Grade (10 points)
* Use the code provided to separate the data by grade (1 points)
* Group by "school_name" and take the mean of each (4 points)
* Use the code to select only the reading_score (1 points)
* Combine each of the scores above into single DataFrame called reading_scores_by_grade (4 points)

In [92]:
# Group by `school_name` and take the mean of the `math_score` column for each grade
ninth_grade_reading_scores = ninth_graders.groupby("school_name")["reading_score"].mean()
tenth_grade_reading_scores = tenth_graders.groupby("school_name")["reading_score"].mean()
eleventh_grade_reading_scores = eleventh_graders.groupby("school_name")["reading_score"].mean()
twelfth_grade_reading_scores = twelfth_graders.groupby("school_name")["reading_score"].mean()

# Combine each of the scores above into a single DataFrame called `math_scores_by_grade`
reading_scores_by_grade = pd.DataFrame({
    "9th Grade": ninth_grade_reading_scores,
    "10th Grade": tenth_grade_reading_scores,
    "11th Grade": eleventh_grade_reading_scores,
    "12th Grade": twelfth_grade_reading_scores
})

# Minor data wrangling
reading_scores_by_grade.index.name = None

# Display the DataFrame
reading_scores_by_grade

9th Grade  10th Grade  11th Grade  12th Grade
Bailey High School     81.303155   80.907183   80.945643   80.912451
Cabrera High School    83.676136   84.253219   83.788382   84.287958
Figueroa High School   81.198598   81.408912   80.640339   81.384863
Ford High School       80.632653   81.262712   80.403642   80.662338
Griffin High School    83.369193   83.706897   84.288089   84.013699
Hernandez High School  80.866860   80.660147   81.396140   80.857143
Holden High School     83.677165   83.324561   83.815534   84.698795
Huang High School      81.290284   81.512386   81.417476   80.305983
Johnson High School    81.260714   80.773431   80.616027   81.227564
Pena High School       83.807273   83.612000   84.335938   84.591160
Rodriguez High School  80.993127   80.629808   80.864811   80.376426
Shelton High School    84.122642   83.441964   84.373786   82.781671
Thomas High School     83.728850   84.254157   83.585542   83.831361
Wilson High School     83.939778   84.021452   83.764608   84.317673
Wright High School     83.833333   83.812757   84.156322   84.073171

## Scores by School Spending
* Use pd.cut with the provided code to bin the data by the spending ranges (2 points)
* Use the code provided to calculate the averages (1 points)
* Create the spending_summary DataFrame using the binned and averaged spending data (2 points)

In [75]:
# Create spending bins
bins = [0, 584, 629, 644, 675]
group_names = ["<$584", "$585-629", "$630-644", "$645-675"]

# Assign spending bins to each row based on budget per student
school_data_complete['spending_bins'] = pd.cut(school_data_complete['budget'] / school_data_complete['size'], bins, labels=group_names)

# Group by spending bins
by_spending = school_data_complete.groupby('spending_bins')

# Calculate average math score, average reading score, and percentage of students passing math, reading, and overall
avg_math_score = by_spending['math_score'].mean()
avg_reading_score = by_spending['reading_score'].mean()
percent_passing_math = (school_data_complete['math_score'] >= 70).groupby(school_data_complete['spending_bins']).mean() * 100
percent_passing_reading = (school_data_complete['reading_score'] >= 70).groupby(school_data_complete['spending_bins']).mean() * 100
percent_overall_passing = ((school_data_complete['math_score'] >= 70) & (school_data_complete['reading_score'] >= 70)).groupby(school_data_complete['spending_bins']).mean() * 100

# Create DataFrame
spending_summary = pd.DataFrame({
    "Average Math Score": avg_math_score,
    "Average Reading Score": avg_reading_score,
    "% Passing Math": percent_passing_math,
    "% Passing Reading": percent_passing_reading,
    "% Overall Passing": percent_overall_passing
})

# Format the DataFrame
spending_summary.index.name = "Spending Ranges (Per Student)"
spending_summary.style.format({
    "Average Math Score": "{:.2f}",
    "Average Reading Score": "{:.2f}",
    "% Passing Math": "{:.2f}",
    "% Passing Reading": "{:.2f}",
    "% Overall Passing": "{:.2f}"
})


## Scores by School Size (5 points)nts)
* Use pd.cut with the provided code to bin the data by the school sizes (2 points)
* Use the code provided to calculate the averages (1 points)
* Create the size_summary DataFrame using the binned and averaged size data (2 points)

In [100]:
# Create bins for segmenting and sorting data
size_bins = [0, 999, 1999, 5000]
school_size_name = ['Small (<1000)', 'Medium (1000-2000)', 'Large (2000-5000)']

# Categorize data by school size
per_school_summary['School Size'] = pd.cut(per_school_summary['Total Students'], bins=size_bins, labels=school_size_name, include_lowest=True)

# Group the dataframe by school size and calculate the average scores
school_size_grouped = per_school_summary.groupby('School Size')
size_summary = school_size_grouped[['Average Math Score', 'Average Reading Score', '% Passing Math', '% Passing Reading', '% Overall Passing']].mean()

# Present the dataframe
size_summary


Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

## Scores by School Type (5 points)
* Group the per_school_summary DataFrame by "School Type" and average the results (2 points)
* Use the code provided to select the new column data (1 point)
* Create a new DataFrame called type_summary that uses the new column data (2 points)

In [103]:
# Select the desired columns for the new DataFrame
type_summary = per_school_summary[["School Type", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]]

# Group the type_summary DataFrame by "School Type" and calculate the average scores
type_summary = type_summary.groupby("School Type").mean()

# Display the type_summary DataFrame
type_summary


Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

# Written Report (15 points)
* Summarizes the analysis (5 points)

* Draws two correct conclusions or comparisons from the calculations (10 points)

### Summarizes the analysis
District Summary:

* Total Schools: There are 15 schools in the district.
* Total Students: The total number of students in all the schools is 39,170.
* Total Budget: The total budget for all the schools is $24,649,428.00.
* Average Math Score: The average math score across all schools is 78.9854.
* Average Reading Score: The average reading score across all schools is 81.8778.
* % Passing Math: 74.9809% of students have a math score of 70 or greater.
* % Passing Reading: 85.8055% of students have a reading score of 70 or greater.
* % Overall Passing: 65.1723% of students passed both math and reading.

### Conclusion
Based on the analysis conducted on the school data, it can be observed that school size, budget and district can have an impact on the passing results of students.

Impact of School Size:
* According to the data, larger schools often have lower passing rates than smaller schools. In smaller schools, there is a tendency for a higher percentage of pupils to pass the math, reading, and overall exams. This shows that smaller schools could offer pupils a more supportive atmosphere for learning and individualized attention, improving academic achievement.

Impact of School Budget:

* According to the data, there is an correlation between school budget and passing rates, with lower budgets all around being associated with greater passing rates in math, reading, and general performance. This suggests that these schools may have employed successful tactics, made efficient use of the resources at their disposal, or prioritized particular areas of instruction that have favorably impacted student progress despite having minimal financial resources.
    

Impact of School District:

* The analysis of school districts indicates that charter schools often do better academically than district schools. Charter schools regularly have higher passing rates in areas like math and reading, as well as overall, according to the results by school type. This implies that charter schools' unique traits and pedagogical strategies could be responsible for their superior performance to district schools. To better understand the causes of this discrepancy and to pinpoint potential best practices that district schools may use to improve their performance, more study is required.